### PR TITLE
#0: fix if sharding width is not multiple of 16 for eltwise unary op.

### PR DESCRIPTION
**Pull request by shwetankTT**

### CODE REVIEW

**Review by AleksKnezevic — APPROVED** [review]

**Review by ntarafdar — APPROVED** [review]

File: tt_eager/tt_dnn/op_library/eltwise_unary/multi_core/eltwise_unary_op_sharded.cpp [hunk 1]
```diff
@@ -35,9 +35,12 @@ operation::ProgramWithCallbacks eltwise_unary_sharded(const Tensor &input, Tenso
     uint32_t output_tile_size = tt::tt_metal::detail::TileSize(out_df);
 
     TT_FATAL(input_tile_size == output_tile_size, "Input and output tile size should be same");
-    uint32_t shard_size_in_bytes = shard_spec.numel() * datum_size(act_df);
+    size_t shard_height = shard_spec.shape[0];
+    size_t shard_width = round_up_to_mul16(shard_spec.shape[1]); // round up to multiple of 16
```
Comment (tt-nshanker) [inline tt_eager/tt_dnn/op_library/eltwise_unary/multi_core/eltwise_unary_op_sharded.cpp:39]: We can't round up the shard width here in the op code. The shard size is used to size the CB which needs to exactly match the l1 size.

**Review by tt-nshanker — CHANGES_REQUESTED** [review]

File: tt_eager/tt_dnn/op_library/eltwise_unary/multi_core/eltwise_unary_op_sharded.cpp (hunk 1 above)
Comment (tt-nshanker) [reply tt_eager/tt_dnn/op_library/eltwise_unary/multi_core/eltwise_unary_op_sharded.cpp:39]: But shard_spec.shape[1] here is not in bytes. So, we can't round that. For example, shard width = 40 is valid for bfloat16 datatype. If we round to 16 here, then shard width would be incorrectly 48. Then, shard_size_in_bytes calculation below this line would have incorrect size of the sharded buffer.

Comment (shwetankTT) [reply tt_eager/tt_dnn/op_library/eltwise_unary/multi_core/eltwise_unary_op_sharded.cpp:39]: That's how the tensor is created. if the shard_width is 40 then page size will be roundto32(40 * 2 ) = 96. Total memory per core for the shard will be 96 * shard_height. 
See https://github.com/tenstorrent/tt-metal/blob/94490875229f70e42fd4004f459afca9484960cd/tt_metal/detail/util.hpp#L30C74-L30C91

Comment (tt-nshanker) [reply tt_eager/tt_dnn/op_library/eltwise_unary/multi_core/eltwise_unary_op_sharded.cpp:39]: Please add comment in the code with this information that shard width is not actual actual shard width.

**Review by tt-nshanker — APPROVED** [review]

**Pull request merged**